pkg/plugins/optional/monitoring/v1alpha/scaffolds/internal/templates/util.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templates
+
+import (
+	"path/filepath"
+
+	"sigs.k8s.io/kubebuilder/v3/pkg/machinery"
+)
+
+var _ machinery.Template = &MetricsUtilManifest{}
+
+// Kustomization scaffolds a file that defines the kustomization scheme for the metrics folder
+type MetricsUtilManifest struct {
+	machinery.TemplateMixin
+}
+
+// SetTemplateDefaults implements file.Template
+func (f *MetricsUtilManifest) SetTemplateDefaults() error {
+	if f.Path == "" {
+		f.Path = filepath.Join("monitoring", "metrics", "util.go")
+	}
+
+	f.TemplateBody = MetricsUtilTemplate
+	f.IfExistsAction = machinery.OverwriteFile
+
+	return nil
+}
+
+// nolint: lll
+const MetricsUtilTemplate = `
+package util
+
+import (
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+type MetricType string
+
+// To add more types visit: https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#pkg-types
+const (
+	Counter MetricType = "counter"
+	Gauge   MetricType = "gauge"
+)
+
+type Metric struct {
+	Name string
+	Help string
+	Type MetricType
+}
+
+var collectors map[string]prometheus.Collector
+
+func RegisterMetrics(allMetrics [][]Metric) {
+	collectors = map[string]prometheus.Collector{}
+
+	for _, m := range flatMetrics(allMetrics) {
+		v := createMetric(m)
+		metrics.Registry.MustRegister(v)
+		collectors[m.Name] = v
+	}
+}
+
+func ListMetrics(allMetrics [][]Metric) []Metric {
+	return flatMetrics(allMetrics)
+}
+
+func GetCounterMetric(metricName string) prometheus.Counter {
+	return castCounter(metricName, getMetric(metricName))
+}
+
```

Comment:
@avlitman I think we should also add here functions for the summery and historgram type

Reply:
If we do that, I think we should add also an example metrics for these types in example_metrics.go as well,
since GetCounterMetric/GetGaugeMetric is used there:
`
func IncrementMetric() {
	util.GetCounterMetric("metric_name").Inc()
}
// Set sets the gauge to an arbitrary value.
func SetStatus(status float64) {
	util.GetGaugeMetric("metric_name").Set(status)
}
`